<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="heinin" />

<meta name="date" content="2024-05-22" />

<title>19130: CAR T Production. Processing 10x data</title>

<script src="site_libs/header-attrs-2.22/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">19130_CAR_production</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">19130: CAR T Production. Processing 10x
data</h1>
<h4 class="author">heinin</h4>
<h4 class="date">2024-05-22</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-05-30
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 5
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 2
</p>
<p>
<strong>Knit directory:</strong> <code>19130_CAR_production/</code>
<span class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted
changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges"
class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown is untracked by Git. To know which version of the R
Markdown file created these results, you’ll want to first commit it to
the Git repo. If you’re still working on the analysis, you can ignore
this warning. When you’re finished, you can run
<code>wflow_publish</code> to commit the R Markdown file and build the
HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20240522code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20240522)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20240522code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20240522)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongabsolute">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>File paths:</strong> absolute </a>
</p>
</div>
<div id="strongFilepathsstrongabsolute" class="panel-collapse collapse">
<div class="panel-body">
<p>
Using absolute paths to the files within your workflowr project makes it
difficult for you and others to run your code on a different machine.
Change the absolute path(s) below to the suggested relative path(s) to
make your code more reproducible.
</p>
<table class="table table-condensed table-hover">
<thead>
<tr>
<th style="text-align:left;">
absolute
</th>
<th style="text-align:left;">
relative
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
/home/hnatri/19130_CAR_production/
</td>
<td style="text-align:left;">
.
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrong68ac815">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong> 68ac815
</a>
</p>
</div>
<div id="strongRepositoryversionstrong68ac815"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version 68ac815.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Untracked files:
    Untracked:  analysis/process_10x.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
There are no past versions. Publish this analysis with
<code>wflow_publish()</code> to start tracking its development.
</p>
<hr>
</div>
</div>
</div>
<div id="processing-10x-data" class="section level2">
<h2>Processing 10x data</h2>
<div id="libraries-and-environment-variables" class="section level3">
<h3>Libraries and environment variables</h3>
<pre class="r"><code>library(workflowr)
library(Seurat)
library(googlesheets4)
library(tidyverse)
library(plyr)
library(SoupX)
library(pheatbuilder)

setwd(&quot;/home/hnatri/19130_CAR_production/&quot;)
set.seed(1234)
options(future.globals.maxSize = 30000 * 1024^2)
reduction &lt;- &quot;umap&quot;

source(&quot;/home/hnatri/SingleCellBestPractices/scripts/preprocessing_qc_module.R&quot;)</code></pre>
</div>
<div id="import-10x-data-demultiplexing" class="section level3">
<h3>Import 10x data, demultiplexing</h3>
<p>The setup is two runs with the same samples:<br />
<em>Run1 (well 3A) F07344-GEX_F07352-CAR_F07348-FB, with a 50-60k pool
load,<br />
</em>Run2 (well 3B) F07345-GEX_F07353-CAR_F07349-FB, with a 90-100k pool
load</p>
<pre class="r"><code># Metadata
# https://docs.google.com/spreadsheets/d/1DstdK4vXjk7ZEeBalW_4CA6Vmd57BGENUx3iMDRLszk/edit?usp=sharing
gs4_deauth()
production_samples  &lt;- gs4_get(&quot;https://docs.google.com/spreadsheets/d/1DstdK4vXjk7ZEeBalW_4CA6Vmd57BGENUx3iMDRLszk/edit?usp=sharing&quot;)
#sheet_names(production_samples)
production_samples &lt;- read_sheet(production_samples, sheet = &quot;Samples&quot;)

# Importing data. Two runs with the same samples.
# Concated GEX: /tgen_labs/banovich/SingleCell/CellRanger/7_1_0/Projects/BCTCSF/HT/BCTCSF_0130_1_PB_CONCAT_C1_X5HT2_F07344-F07345_CONCAT
path_list &lt;- list(&quot;run1&quot; = unique(production_samples[which(production_samples$Run==&quot;run1&quot;),]$CellRanger_path),
                  &quot;run2&quot; = unique(production_samples[which(production_samples$Run==&quot;run2&quot;),]$CellRanger_path))

# Cell hashing antibody names
hash_antibodies &lt;- unique(production_samples$Hash_ID)

# Demultiplexing
seurat_list &lt;- prep_seurat_list_multiplexed(metadata = production_samples,
                                            batch_ID = &quot;Run&quot;,
                                            cellRanger_path = &quot;CellRanger_path&quot;,
                                            cell_ID_prefix = &quot;Sample_Name&quot;,
                                            CellHashing_Ab = &quot;Hash_ID&quot;,
                                            Hash_Abs = hash_antibodies)</code></pre>
<p>Singlets and doublets for each run.</p>
<pre class="r"><code># Singlets/doublets
table(seurat_list[[&quot;run1&quot;]]$Hash_classification.global)</code></pre>
<pre><code>
 Doublet Negative  Singlet 
   21952    19725      603 </code></pre>
<pre class="r"><code>table(seurat_list[[&quot;run2&quot;]]$Hash_classification.global)</code></pre>
<pre><code>
 Doublet Negative  Singlet 
    7543    37120     7914 </code></pre>
<pre class="r"><code>p1 &lt;- table(seurat_list[[&quot;run1&quot;]]$Hash_classification.global) %&gt;% as.data.frame() %&gt;%
  ggplot(aes(x = Var1, y = Freq, fill = Var1)) +
  geom_bar(position=&quot;dodge&quot;, stat=&quot;identity&quot;) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  xlab(&quot;Hash classification for run 1&quot;) +
  ylab(&quot;# cells&quot;) +
  NoLegend()

p2 &lt;- table(seurat_list[[&quot;run1&quot;]]$Hash_maxID) %&gt;% as.data.frame() %&gt;%
  ggplot(aes(x = Var1, y = Freq, fill = Var1)) +
  geom_bar(position=&quot;dodge&quot;, stat=&quot;identity&quot;) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  xlab(&quot;Hash classification for run 1&quot;) +
  ylab(&quot;# cells&quot;) +
  NoLegend()

p3 &lt;- table(seurat_list[[&quot;run2&quot;]]$Hash_classification.global) %&gt;% as.data.frame() %&gt;%
  ggplot(aes(x = Var1, y = Freq, fill = Var1)) +
  geom_bar(position=&quot;dodge&quot;, stat=&quot;identity&quot;) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  xlab(&quot;Hash classification for run 2&quot;) +
  ylab(&quot;# cells&quot;) +
  NoLegend()

p4 &lt;- table(seurat_list[[&quot;run2&quot;]]$Hash_maxID) %&gt;% as.data.frame() %&gt;%
  ggplot(aes(x = Var1, y = Freq, fill = Var1)) +
  geom_bar(position=&quot;dodge&quot;, stat=&quot;identity&quot;) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  xlab(&quot;Hash classification for run 2&quot;) +
  ylab(&quot;# cells&quot;) +
  NoLegend()

(p1 + p2) / (p3 + p4)</code></pre>
<p><img src="figure/process_10x.Rmd/unnamed-chunk-3-1.png" width="1152" style="display: block; margin: auto;" /></p>
<p>Plotting HTO counts.</p>
<pre class="r"><code># Plotting HTO counts
hash_counts &lt;- lapply(seurat_list, function(xx){
  hash &lt;- LayerData(xx, assay = &quot;Hash&quot;, layer = &quot;counts&quot;)
  hash &lt;- as.data.frame(t(hash))
  hash$hash_classification &lt;- mapvalues(rownames(hash),
                             from = rownames(xx@meta.data),
                             to = xx@meta.data$Hash_maxID)
  
  hash
})

hash_antibodies &lt;- gsub(&quot;_&quot;, &quot;-&quot;, hash_antibodies)</code></pre>
<p>Run 1</p>
<pre class="r"><code>hash_counts[[1]] %&gt;% pivot_longer(cols = hash_antibodies,
                                  values_to = &quot;counts&quot;,
                                  names_to = &quot;HashAB&quot;) %&gt;%
  group_by(HashAB) %&gt;%
  dplyr::summarize(sum = sum(counts)) %&gt;%
  ungroup() %&gt;%
  ggplot(aes(x = HashAB, y = sum, fill = HashAB)) +
  geom_bar(position=&quot;dodge&quot;, stat=&quot;identity&quot;) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  xlab(&quot;Hash counts in run 1&quot;) +
  ylab(&quot;# cells&quot;)</code></pre>
<p><img src="figure/process_10x.Rmd/unnamed-chunk-5-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>hash_counts[[1]] %&gt;% pivot_longer(cols = hash_antibodies,
                                  values_to = &quot;counts&quot;,
                                  names_to = &quot;HashAB&quot;) %&gt;%
  group_by(hash_classification, HashAB) %&gt;%
  dplyr::summarize(sum = sum(counts)) %&gt;%
  ungroup() %&gt;%
  pivot_wider(values_from = sum,
              names_from = HashAB) %&gt;%
  pheat()</code></pre>
<p><img src="figure/process_10x.Rmd/unnamed-chunk-5-2.png" width="672" style="display: block; margin: auto;" /></p>
<p>Run 2</p>
<pre class="r"><code>hash_counts[[2]] %&gt;% pivot_longer(cols = hash_antibodies,
                                  values_to = &quot;counts&quot;,
                                  names_to = &quot;HashAB&quot;) %&gt;%
  group_by(HashAB) %&gt;%
  dplyr::summarize(sum = sum(counts)) %&gt;%
  ungroup() %&gt;%
  ggplot(aes(x = HashAB, y = sum, fill = HashAB)) +
  geom_bar(position=&quot;dodge&quot;, stat=&quot;identity&quot;) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  xlab(&quot;Hash counts in run 2&quot;) +
  ylab(&quot;# cells&quot;)</code></pre>
<p><img src="figure/process_10x.Rmd/unnamed-chunk-6-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>hash_counts[[2]] %&gt;% pivot_longer(cols = hash_antibodies,
                                  values_to = &quot;counts&quot;,
                                  names_to = &quot;HashAB&quot;) %&gt;%
  group_by(hash_classification, HashAB) %&gt;%
  dplyr::summarize(sum = sum(counts)) %&gt;%
  ungroup() %&gt;%
  pivot_wider(values_from = sum,
              names_from = HashAB) %&gt;%
  pheat()</code></pre>
<p><img src="figure/process_10x.Rmd/unnamed-chunk-6-2.png" width="672" style="display: block; margin: auto;" /></p>
<p>Singlets and doublets in the combined data.</p>
<pre class="r"><code># Merging for plotting
seurat_merged &lt;- merge(x = seurat_list[[1]], y = seurat_list[[2]])

# Singlets/doublets
table(seurat_merged$Hash_classification.global) %&gt;% as.data.frame() %&gt;%
  ggplot(aes(x = Var1, y = Freq, fill = Var1)) +
  geom_bar(position=&quot;dodge&quot;, stat=&quot;identity&quot;) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  xlab(&quot;Hash classification for both runs&quot;) +
  ylab(&quot;# cells&quot;) +
  NoLegend()</code></pre>
<p><img src="figure/process_10x.Rmd/unnamed-chunk-7-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>table(seurat_merged$Hash_maxID) %&gt;% as.data.frame() %&gt;%
  ggplot(aes(x = Var1, y = Freq, fill = Var1)) +
  geom_bar(position=&quot;dodge&quot;, stat=&quot;identity&quot;) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  xlab(&quot;Hash classification for both runs&quot;) +
  ylab(&quot;# cells&quot;) +
  NoLegend()</code></pre>
<p><img src="figure/process_10x.Rmd/unnamed-chunk-7-2.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>seurat_list &lt;- lapply(seurat_list, function(xx){
  xx@meta.data$cellname &lt;- colnames(xx)
  
  xx
})</code></pre>
<p>Filtering to retain singlets</p>
<pre class="r"><code># Filtering to retain singlets
seurat_singlets_list &lt;- lapply(seurat_list, function(xx){
  xx &lt;- subset(xx, subset = Hash_classification.global == &quot;Singlet&quot;)
  
  xx
})
names(seurat_singlets_list) &lt;- names(seurat_list)

# Merging for plotting
seurat_merged &lt;- merge(x = seurat_list[[1]], y = seurat_list[[2]])

# Plotting
table(seurat_merged$Sample_Name) %&gt;% as.data.frame() %&gt;%
  ggplot(aes(x = Var1, y = Freq, fill = Var1)) +
  geom_bar(position=&quot;dodge&quot;, stat=&quot;identity&quot;) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  xlab(&quot;Sample&quot;) +
  ylab(&quot;# singets&quot;)</code></pre>
<p><img src="figure/process_10x.Rmd/unnamed-chunk-8-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>table(seurat_merged$Run) %&gt;% as.data.frame() %&gt;%
  ggplot(aes(x = Var1, y = Freq, fill = Var1)) +
  geom_bar(position=&quot;dodge&quot;, stat=&quot;identity&quot;) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  xlab(&quot;Run&quot;) +
  ylab(&quot;# singlets&quot;)</code></pre>
<p><img src="figure/process_10x.Rmd/unnamed-chunk-8-2.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="soupx-for-ambient-rna" class="section level3">
<h3>SoupX for ambient RNA</h3>
<pre class="r"><code># Saving demultiplexed data in 10x format
for (i in 1:length(seurat_singlets_list)){
  print(paste0(&quot;Converting sample &quot;, names(seurat_singlets_list[i])))
  obj.sub &lt;- seurat_singlets_list[[i]] 
  
  counts &lt;- LayerData(obj.sub, assay = &quot;RNA&quot;, layer = &quot;counts&quot;)
  DropletUtils::write10xCounts(path = paste0(&quot;/scratch/hnatri/CART/SoupX/demultiplexed_&quot;, names(seurat_singlets_list[i])), 
                               x = counts, 
                               barcodes = colnames(counts), # cell names
                               gene.id = rownames(counts), # Gene identifiers, one per row of X
                               type = &quot;sparse&quot;, version=&quot;3&quot;, overwrite = T)
}</code></pre>
<pre><code>[1] &quot;Converting sample run1&quot;
[1] &quot;Converting sample run2&quot;</code></pre>
<pre class="r"><code># Running SoupX
seurat_soupx_list &lt;- sapply(names(seurat_singlets_list), function(xx){
  print(xx)

  # Read in count and droplet data
  # Converted demultiplexed counts
  message(&quot;Reading converted demultiplexed counts&quot;)
  d10x_toc &lt;- Read10X(paste0(&quot;/scratch/hnatri/CART/SoupX/demultiplexed_&quot;, xx))
  
  #batch_id &lt;- seurat_singlets_list[[xx]]$Run
  
  # Need to read in batch specific empty droplet file
  message(&quot;Reading empty droplet data&quot;)
  d10x_tod &lt;- Read10X(paste0(path_list[[xx]], &quot;/outs/raw_feature_bc_matrix/&quot;))
  
  if(length(names(d10x_tod))&gt;1){
    d10x_tod_gex &lt;- d10x_tod[[1]]
  } else {
    d10x_tod_gex &lt;- d10x_tod
  }
  
  colnames(d10x_tod_gex) &lt;- paste0(xx, &quot;_&quot;, colnames(d10x_tod_gex))
  
  # Some batches only have features that are expressed in at least one cell;
  # need to fix feature order
  #rownames(d10x_toc) &lt;- gsub(&quot;_&quot;, &quot;-&quot;, rownames(d10x_toc))
  #rownames(d10x_tod_gex) &lt;- gsub(&quot;_&quot;, &quot;-&quot;, rownames(d10x_tod_gex))
  d10x_tod_gex &lt;- d10x_tod_gex[rownames(d10x_toc),]
  
  # Run SoupX
  sc &lt;- SoupChannel(d10x_tod_gex, d10x_toc, calcSoupProfile = FALSE) 
  
  sc &lt;- estimateSoup(sc)
  toc_seu &lt;- CreateSeuratObject(d10x_toc)
  toc_seu &lt;- SCTransform(toc_seu, vst.flavor = &quot;v2&quot;)
  if(ncol(toc_seu)&lt;50){
    toc_seu &lt;- RunPCA(toc_seu, npcs = ncol(toc_seu)-1)
  } else{
    toc_seu &lt;- RunPCA(toc_seu)
  }
  toc_seu &lt;- RunUMAP(toc_seu, dims = 1:15)
  toc_seu &lt;- FindNeighbors(toc_seu, dims = 1:15)
  toc_seu &lt;- FindClusters(toc_seu, resolution = 1)

  # Add meta data to soupX object
  sc &lt;- setClusters(sc, setNames(toc_seu$seurat_clusters, rownames(toc_seu@meta.data)))
  # Estimate contamination (automated method)
  message(paste0(&quot;Getting autoEstCont for: &quot;, i))
  sc &lt;- autoEstCont(sc, tfidfMin = 0.6, soupQuantile = 0.7, forceAccept = TRUE, doPlot = F) 
  out &lt;- adjustCounts(sc)
  
  # Create Seurat object using corrected data
  d10x_seu &lt;- CreateSeuratObject(out, assay = &quot;SoupX_RNA&quot;)
  d10x_seu[[&quot;RNA&quot;]] &lt;- toc_seu@assays[[&quot;RNA&quot;]]
  d10x_seu &lt;- PercentageFeatureSet(d10x_seu, pattern = &quot;^MT-&quot;, col.name = &quot;percent.mt_RNA&quot;, assay = &quot;RNA&quot;)
  d10x_seu &lt;- PercentageFeatureSet(d10x_seu, pattern = &quot;^RP[SL]|^MRP[SL]&quot;, col.name = &quot;percent.ribo_RNA&quot;, assay = &quot;RNA&quot;)
  d10x_seu &lt;- PercentageFeatureSet(d10x_seu, pattern = &quot;^MT-&quot;, col.name = &quot;percent.mt_SoupX_RNA&quot;, assay = &quot;SoupX_RNA&quot;)
  d10x_seu &lt;- PercentageFeatureSet(d10x_seu, pattern = &quot;^RP[SL]|^MRP[SL]&quot;, col.name = &quot;percent.ribo_SoupX_RNA&quot;, assay = &quot;SoupX_RNA&quot;)
  
  # Add sample metadata
  d10x_seu$Sample &lt;- paste0(sapply(strsplit(colnames(d10x_seu), &quot;_&quot;), `[`, 1), &quot;_&quot;,
                            sapply(strsplit(colnames(d10x_seu), &quot;_&quot;), `[`, 2))
  

  d10x_seu$Day &lt;-paste0(sapply(strsplit(colnames(d10x_seu), &quot;_&quot;), `[`, 2))
  d10x_seu$Sample_Type &lt;- paste0(sapply(strsplit(colnames(d10x_seu), &quot;_&quot;), `[`, 1))
  d10x_seu$cellname &lt;- colnames(d10x_seu)
  d10x_seu$Run &lt;- xx
  
  d10x_seu &lt;- RenameCells(d10x_seu,
                          new.names = paste0(xx,&quot;_&quot;, colnames(d10x_seu)))
  
  d10x_seu
})</code></pre>
<pre><code>[1] &quot;run1&quot;
Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 603
Number of edges: 23686

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.6416
Number of communities: 7
Elapsed time: 0 seconds
[1] &quot;run2&quot;
Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 7914
Number of edges: 262762

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.8566
Number of communities: 19
Elapsed time: 0 seconds</code></pre>
<pre class="r"><code>names(seurat_soupx_list) &lt;- names(seurat_singlets_list)

# Merge
seurat_merged &lt;- merge(x = seurat_soupx_list[[1]], y = seurat_soupx_list[[2]])

#saveRDS(seurat_merged, &quot;/scratch/hnatri/CART/SoupX/seurat_merged.rds&quot;)</code></pre>
</div>
<div id="qc-and-filtering" class="section level3">
<h3>QC and filtering</h3>
<pre class="r"><code>par(mfrow=c(2,3))

# PLOTS 1 &amp; 2: nCount vs. nFeature
smoothScatter(log2(seurat_merged$nCount_SoupX_RNA), log2(seurat_merged$nCount_SoupX_RNA),
              xlab = &quot;log2(nCount_SoupX_RNA)&quot;, ylab = &quot;log2(nFeature_SoupX_RNA)&quot;)

smoothScatter(seurat_merged$nCount_SoupX_RNA, seurat_merged$nCount_SoupX_RNA,
              xlab = &quot;nCount_SoupX_RNA&quot;, ylab = &quot;nFeature_SoupX_RNA&quot;)

# PLOTS 3 &amp; 4: nCount vs. percent.mt_RNA
smoothScatter(seurat_merged$percent.mt_SoupX_RNA, log2(seurat_merged$nCount_SoupX_RNA),
              xlab = &quot;% MT&quot;, ylab = &quot;log2(nCount_SoupX_RNA)&quot;)

smoothScatter(seurat_merged$percent.mt_SoupX_RNA, seurat_merged$nCount_SoupX_RNA,
              xlab = &quot;% MT&quot;, ylab = &quot;nCount_SoupX_RNA&quot;)
abline(v = 10, h = 1000, 
       lty = &quot;dashed&quot;, lwd = 1.25, col = &quot;red&quot;)

# PLOTS 5 &amp; 6: nFeature vs. percent.mt_RNA
smoothScatter(seurat_merged$percent.mt_SoupX_RNA, log2(seurat_merged$nFeature_SoupX_RNA),
              xlab = &quot;% MT&quot;, ylab = &quot;log2(nFeature_RNA)&quot;)

smoothScatter(seurat_merged$percent.mt_SoupX_RNA, seurat_merged$nFeature_SoupX_RNA,
              xlab = &quot;% MT&quot;, ylab = &quot;nFeature_RNA&quot;)
abline(v = 10, h = 500, 
       lty = &quot;dashed&quot;, lwd = 1.25, col = &quot;red&quot;)</code></pre>
<p><img src="figure/process_10x.Rmd/unnamed-chunk-10-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>seurat_filtered &lt;- subset(seurat_merged, subset = percent.mt_SoupX_RNA &lt; 10 &amp;
                            nFeature_SoupX_RNA &gt; 500 &amp; nCount_SoupX_RNA &gt;1000)

saveRDS(seurat_filtered, &quot;/scratch/hnatri/CART/19130_CARprod_filtered.rds&quot;)</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.3.0 (2023-04-21)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] scDblFinder_1.16.0          DoubletFinder_2.0.3        
 [3] scater_1.30.1               scuttle_1.12.0             
 [5] SingleCellExperiment_1.24.0 SummarizedExperiment_1.32.0
 [7] Biobase_2.62.0              GenomicRanges_1.54.1       
 [9] GenomeInfoDb_1.38.8         IRanges_2.36.0             
[11] S4Vectors_0.40.2            BiocGenerics_0.48.1        
[13] MatrixGenerics_1.14.0       matrixStats_1.0.0          
[15] pheatbuilder_0.1.0          SoupX_1.6.2                
[17] plyr_1.8.8                  lubridate_1.9.2            
[19] forcats_1.0.0               stringr_1.5.0              
[21] dplyr_1.1.2                 purrr_1.0.2                
[23] readr_2.1.4                 tidyr_1.3.0                
[25] tibble_3.2.1                ggplot2_3.4.2              
[27] tidyverse_2.0.0             googlesheets4_1.1.0        
[29] Seurat_5.0.1                SeuratObject_5.0.1         
[31] sp_1.6-1                    workflowr_1.7.1            

loaded via a namespace (and not attached):
  [1] fs_1.6.2                  spatstat.sparse_3.0-1    
  [3] bitops_1.0-7              httr_1.4.6               
  [5] RColorBrewer_1.1-3        tools_4.3.0              
  [7] sctransform_0.4.1         utf8_1.2.3               
  [9] R6_2.5.1                  HDF5Array_1.30.1         
 [11] lazyeval_0.2.2            uwot_0.1.14              
 [13] rhdf5filters_1.14.1       withr_2.5.0              
 [15] gridExtra_2.3             progressr_0.13.0         
 [17] cli_3.6.1                 spatstat.explore_3.2-1   
 [19] fastDummies_1.7.3         labeling_0.4.2           
 [21] sass_0.4.6                spatstat.data_3.0-1      
 [23] ggridges_0.5.4            pbapply_1.7-0            
 [25] Rsamtools_2.18.0          R.utils_2.12.2           
 [27] parallelly_1.36.0         limma_3.58.1             
 [29] rstudioapi_0.14           generics_0.1.3           
 [31] BiocIO_1.12.0             ica_1.0-3                
 [33] spatstat.random_3.1-5     Matrix_1.6-5             
 [35] ggbeeswarm_0.7.2          fansi_1.0.4              
 [37] abind_1.4-5               R.methodsS3_1.8.2        
 [39] lifecycle_1.0.3           whisker_0.4.1            
 [41] yaml_2.3.7                edgeR_4.0.16             
 [43] glmGamPoi_1.14.3          rhdf5_2.46.1             
 [45] SparseArray_1.2.4         Rtsne_0.16               
 [47] grid_4.3.0                promises_1.2.0.1         
 [49] dqrng_0.3.0               crayon_1.5.2             
 [51] miniUI_0.1.1.1            lattice_0.21-8           
 [53] beachmat_2.18.1           cowplot_1.1.1            
 [55] metapod_1.10.1            pillar_1.9.0             
 [57] knitr_1.43                rjson_0.2.21             
 [59] xgboost_1.7.5.1           future.apply_1.11.0      
 [61] codetools_0.2-19          leiden_0.4.3             
 [63] glue_1.6.2                getPass_0.2-4            
 [65] data.table_1.14.8         vctrs_0.6.5              
 [67] png_0.1-8                 spam_2.9-1               
 [69] cellranger_1.1.0          gtable_0.3.3             
 [71] cachem_1.0.8              xfun_0.39                
 [73] DropletUtils_1.22.0       S4Arrays_1.2.1           
 [75] mime_0.12                 survival_3.5-5           
 [77] gargle_1.4.0              pheatmap_1.0.12          
 [79] statmod_1.5.0             bluster_1.12.0           
 [81] ellipsis_0.3.2            fitdistrplus_1.1-11      
 [83] ROCR_1.0-11               nlme_3.1-162             
 [85] RcppAnnoy_0.0.20          rprojroot_2.0.3          
 [87] bslib_0.4.2               irlba_2.3.5.1            
 [89] vipor_0.4.5               KernSmooth_2.23-21       
 [91] colorspace_2.1-0          tidyselect_1.2.0         
 [93] processx_3.8.1            curl_5.0.1               
 [95] compiler_4.3.0            git2r_0.32.0             
 [97] BiocNeighbors_1.20.2      DelayedArray_0.28.0      
 [99] plotly_4.10.2             rtracklayer_1.62.0       
[101] scales_1.2.1              lmtest_0.9-40            
[103] callr_3.7.3               digest_0.6.31            
[105] goftest_1.2-3             spatstat.utils_3.0-3     
[107] rmarkdown_2.22            XVector_0.42.0           
[109] htmltools_0.5.5           pkgconfig_2.0.3          
[111] sparseMatrixStats_1.14.0  highr_0.10               
[113] fastmap_1.1.1             rlang_1.1.1              
[115] htmlwidgets_1.6.2         shiny_1.7.4              
[117] DelayedMatrixStats_1.24.0 farver_2.1.1             
[119] jquerylib_0.1.4           zoo_1.8-12               
[121] jsonlite_1.8.5            BiocParallel_1.36.0      
[123] R.oo_1.25.0               BiocSingular_1.18.0      
[125] RCurl_1.98-1.12           magrittr_2.0.3           
[127] GenomeInfoDbData_1.2.11   dotCall64_1.0-2          
[129] patchwork_1.1.2           Rhdf5lib_1.24.1          
[131] munsell_0.5.0             Rcpp_1.0.10              
[133] viridis_0.6.3             reticulate_1.29          
[135] stringi_1.7.12            zlibbioc_1.48.0          
[137] MASS_7.3-60               parallel_4.3.0           
[139] listenv_0.9.0             ggrepel_0.9.3            
[141] deldir_1.0-9              Biostrings_2.70.3        
[143] splines_4.3.0             tensor_1.5               
[145] hms_1.1.3                 locfit_1.5-9.8           
[147] ps_1.7.5                  igraph_1.4.3             
[149] spatstat.geom_3.2-1       RcppHNSW_0.5.0           
[151] reshape2_1.4.4            ScaledMatrix_1.10.0      
[153] XML_3.99-0.14             evaluate_0.21            
[155] scran_1.30.2              tzdb_0.4.0               
[157] httpuv_1.6.11             RANN_2.6.1               
[159] polyclip_1.10-4           future_1.32.0            
[161] scattermore_1.2           rsvd_1.0.5               
[163] xtable_1.8-4              restfulr_0.0.15          
[165] RSpectra_0.16-1           later_1.3.1              
[167] viridisLite_0.4.2         googledrive_2.1.0        
[169] beeswarm_0.4.0            GenomicAlignments_1.38.2 
[171] cluster_2.1.4             timechange_0.2.0         
[173] globals_0.16.2           </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
